### PR TITLE
Serialization 2.0

### PR DESCRIPTION
**Pull request by hamishun**
- Breaking backward compatibility changes
- Major Context refactor with API changes
- New serializatipn API
- New binary IO format
- JSON for text format
- New tutorials

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,6 +14,11 @@ cmake_minimum_required(VERSION 3.10.2 FATAL_ERROR)
 
 project(helib_superbuild LANGUAGES C CXX)
 
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline CMakeLists.txt:17]:
```suggestion
# STRINGS avoids having the 2 newline characters at the end of the string.
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -14,6 +14,11 @@ cmake_minimum_required(VERSION 3.10.2 FATAL_ERROR)
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline CMakeLists.txt:18]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: README.md [hunk 3]
```diff
@@ -28,6 +28,9 @@ See [CHANGES.md](CHANGES.md) for more details.
 Full installation instructions and a list of the required dependencies can be found 
 in [INSTALL.md](INSTALL.md).
 
+For guidance in getting started programming with HElib, take a look at the
+example programs and our CKKS tutorial located in `examples` directory.
```
Comment (jlhcrawford) [inline README.md:32]:
```suggestion
example programs and our CKKS tutorial located in the `examples` directory.
```

File: benchmarks/CMakeLists.txt [hunk 4]
```diff
@@ -47,14 +47,20 @@ endif (NOT CMAKE_BUILD_TYPE)
 set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS
     "Debug" "RelWithDebInfo" "Release" "MinSizeRel")
 
-find_package(helib 1.3.0 EXACT REQUIRED)
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline benchmarks/CMakeLists.txt:50]:
```suggestion
# STRINGS avoids having the 2 newline characters at the end of the string.
```

File: benchmarks/CMakeLists.txt [hunk 5]
```diff
@@ -47,14 +47,20 @@ endif (NOT CMAKE_BUILD_TYPE)
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline benchmarks/CMakeLists.txt:51]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: benchmarks/IO.cpp [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+#include <benchmark/benchmark.h>
+#include <iostream>
+
+#include <helib/helib.h>
+#include <helib/debugging.h>
+#include "bgv_common.h"
+
+namespace {
+
+// NOTE: The API in this benchmark have been disabled.
```
Comment (jlhcrawford) [inline benchmarks/IO.cpp:22]: I need to have a look at this and see if should be deleted or replaced with a different benchmark.

File: examples/CMakeLists.txt [hunk 7]
```diff
@@ -31,8 +31,15 @@ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY
     "${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_BINDIR}")
 
-find_package(helib 1.3.0 EXACT REQUIRED)
+
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline examples/CMakeLists.txt:35]:
```suggestion
# STRINGS avoids having the 2 newline characters at the end of the string.
```

File: examples/CMakeLists.txt [hunk 8]
```diff
@@ -31,8 +31,15 @@ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline examples/CMakeLists.txt:36]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: include/helib/Context.h [hunk 9]
```diff
@@ -131,23 +94,115 @@ class ContextBuilder;
...
-  // Forward declarations of useful param structs for Context and
-  // ContextBuilder.
+  // Forward declarations of useful param structs
+  // for Context and ContextBuilder.
   struct ModChainParams;
   struct BootStrapParams;
 
-  std::vector<Cmodulus> moduli; // Cmodulus objects for the different primes
-  // This is private since the implementation assumes that the list of
+  // For serialisation.
```
Comment (jlhcrawford) [inline include/helib/Context.h:110]:
```suggestion
  // For serialization.
```

File: include/helib/Context.h [hunk 10]
```diff
@@ -299,24 +468,28 @@ class Context
   // We then multiply the sqrt of the variance by scale to get
   // the high probability bound.
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:472]: TODO

File: include/helib/Context.h [hunk 11]
```diff
@@ -299,24 +468,28 @@ class Context
...
+   **/
   double noiseBoundForUniform(double magBound, long degBound) const
   {
     return scale * std::sqrt(double(degBound) / 3.0) * magBound;
   }
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:480]: TODO

File: include/helib/Context.h [hunk 12]
```diff
@@ -337,6 +510,9 @@ class Context
   // NOTE: this is slightly more accurate that just calling
   // noiseBoundForUniform with magBound=modulus/2.
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:514]: TODO

File: include/helib/Context.h [hunk 13]
```diff
@@ -346,38 +522,40 @@ class Context
...
+  // of the canonical embedding exceeds B with probability at most
+  // epsilon.
 
   // NOTE: if we evaluate f at a primitive root of unity,
   // then we get a normal random variable variance degBound * sigma^2.
   // We then multiply the sqrt of the variance by scale to get
   // the high probability bound.
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:538]: TODO

File: include/helib/Context.h [hunk 14]
```diff
@@ -346,38 +522,40 @@ class Context
...
+   **/
   double noiseBoundForGaussian(double sigma, long degBound) const
   {
     return scale * std::sqrt(double(degBound)) * sigma;
   }
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:546]: TODO

File: include/helib/Context.h [hunk 15]
```diff
@@ -387,18 +565,19 @@ class Context
   // We then multiply the sqrt of the variance by scale to get
   // the high probability bound.
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:569]: TODO

File: include/helib/Context.h [hunk 16]
```diff
@@ -410,31 +589,35 @@ class Context
   // NOTE: degBound is not used here, but I include it
   // for consistency with the other noiseBound routines
 
+  /**
+   *
```
Comment (jlhcrawford) [inline include/helib/Context.h:593]: TODO

File: misc/psi/CMakeLists.txt [hunk 17]
```diff
@@ -31,7 +31,12 @@ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY
     "${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_BINDIR}")
 
-  find_package(helib 1.3.0 EXACT REQUIRED)
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline misc/psi/CMakeLists.txt:34]:
```suggestion
# STRINGS avoids having the 2 newline characters at the end of the string.
```

File: misc/psi/CMakeLists.txt [hunk 18]
```diff
@@ -31,7 +31,12 @@ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline misc/psi/CMakeLists.txt:35]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: misc/psi/lookup/CMakeLists.txt [hunk 19]
```diff
@@ -20,7 +20,12 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 project(lookup LANGUAGES CXX)
 
-find_package(helib 1.3.0 EXACT REQUIRED)
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline misc/psi/lookup/CMakeLists.txt:23]:
```suggestion
# STRINGS avoids having the 2 characters newline at the end of the string.
```

File: misc/psi/lookup/CMakeLists.txt [hunk 20]
```diff
@@ -20,7 +20,12 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline misc/psi/lookup/CMakeLists.txt:24]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: misc/psi/scoring/CMakeLists.txt [hunk 21]
```diff
@@ -20,7 +20,12 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
 project(scoring LANGUAGES CXX)
 
-find_package(helib 1.3.0 EXACT REQUIRED)
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline misc/psi/scoring/CMakeLists.txt:23]:
```suggestion
# STRINGS avoids having the 2 newline characters at the end of the string.
```

File: misc/psi/scoring/CMakeLists.txt [hunk 22]
```diff
@@ -20,7 +20,12 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline misc/psi/scoring/CMakeLists.txt:24]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: src/Context.cpp [hunk 23]
```diff
@@ -679,27 +765,445 @@ Context::Context(long m,
...
+void Context::addSpecialPrime(long q)
+{
+  assertFalse(inChain(q), "Special prime q is already in the prime chain");
+  long i = moduli.size(); // The index of the new prime in the list
+  moduli.push_back(Cmodulus(zMStar, q, 0));
+  specialPrimes.insert(i);
+}
+
+// Determine the target size of the ctxtPrimes. The target size is
+// set at 2^n, where n is at most NTL_SP_NBITS and at least least
```
Comment (jlhcrawford) [inline src/Context.cpp:900]:
```suggestion
// set at 2^n, where n is at most NTL_SP_NBITS and at least
```

File: src/Context.cpp [hunk 24]
```diff
@@ -679,27 +765,445 @@ Context::Context(long m,
...
+    else
+      h = getHwt();
+
+    double log_phim = std::log(phim);
+    if (log_phim < 1)
+      log_phim = 1;
+
+    if (isCKKS()) {
+      // This is based on a smaller noise estimate so as
+      // to better protext precision...this is based on
```
Comment (jlhcrawford) [inline src/Context.cpp:1051]:
```suggestion
      // to better protect precision...this is based on
```

File: src/Context.cpp [hunk 25]
```diff
@@ -717,48 +1221,45 @@ Context* ContextBuilder<SCHEME>::buildPtr() const
   return new Context(m_, p_, r_, gens_, ords_, mparams, bparams);
 }
 
+// Essentially serialisation of params.
```
Comment (jlhcrawford) [inline src/Context.cpp:1224]:
```suggestion
// Essentially serialization of params.
```

File: src/io.h [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+inline T executeRedirectJsonError(const TCALL& f)
+{
+  try {
+    return f();
+  } catch (const nlohmann::detail::exception& e) {
+    throw IOError(std::string("Error with JSON IO. ") + e.what());
+  }
+}
+
+// template <typename T>
```
Comment (jlhcrawford) [inline src/io.h:208]: Dead code? Should we remove this?

File: src/keySwitching.cpp [hunk 27]
```diff
@@ -229,24 +215,83 @@ void KeySwitch::write(std::ostream& str) const
...
+   */
+  json j = {{"fromKey", unwrap(this->fromKey.writeToJSON())},
+            {"toKeyID", this->toKeyID},
+            {"ptxtSpace", this->ptxtSpace},
+            {"b", writeVectorToJSON(b)},
+            {"prgSeed", prgSeed},
+            {"noiseBound", noiseBound}};
+
+  return wrap(toTypedJson<KeySwitch>(j));
+} // namespace helib
```
Comment (jlhcrawford) [inline src/keySwitching.cpp:262]:
```suggestion
}
```

File: src/keys.cpp [hunk 28]
```diff
@@ -79,11 +81,11 @@ double RLWE(DoubleCRT& c0,
 /********************************************************************/
 // Computes the keySwitchMap pointers, using breadth-first search (BFS)
 
-PubKey::PubKey() :
-    context(*activeContext), pubEncrKey(*this), recryptEkey(*this)
-{
-  recryptKeyID = -1;
-}
+// PubKey::PubKey() :
```
Comment (jlhcrawford) [inline src/keys.cpp:84]: Dead code? Can this be removed?

File: src/keys.cpp [hunk 29]
```diff
@@ -974,65 +899,169 @@ void writePubKeyBinary(std::ostream& str, const PubKey& pk)
...
+
+void PubKey::readJSON(const JsonWrapper& tjw)
+{
+  auto body = [&, this]() {
+    json j = fromTypedJson<PubKey>(unwrap(tjw));
+    Context ser_context = Context::readFromJSON(wrap(j.at("context")));
+    assertEq(context, ser_context, "Context mismatch");
+
+    this->clear();
+    //  std::cerr << "PubKey[";
```
Comment (jlhcrawford) [inline src/keys.cpp:1029]: Can this line be removed?

File: src/primeChain.cpp [hunk 30]
```diff
@@ -19,45 +19,57 @@
...
+#include "io.h"
+
 namespace helib {
 
 inline bool operator>(const ModuliSizes::Entry& a, const ModuliSizes::Entry& b)
 {
   return a.first > b.first;
 }
 
+// FIXME: ADD JSON SERIALIZATION
```
Comment (jlhcrawford) [inline src/primeChain.cpp:35]: Is this comment still relevant?

File: utils/CMakeLists.txt [hunk 31]
```diff
@@ -31,7 +31,12 @@ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY
     "${CMAKE_BINARY_DIR}/${CMAKE_INSTALL_BINDIR}")
 
-find_package(helib 1.3.0 EXACT REQUIRED)
+# STRINGS avoids having the 2 weird newline at the end of the string
```
Comment (jlhcrawford) [inline utils/CMakeLists.txt:34]:
```suggestion
# STRINGS avoids having the 2 newline characters at the end of the string.
```

File: utils/CMakeLists.txt [hunk 32]
```diff
@@ -31,7 +31,12 @@ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY
...
+# Alternatively it's possible to use file(READ and then
```
Comment (jlhcrawford) [inline utils/CMakeLists.txt:35]:
```suggestion
# Alternatively it's possible to use file(READ ...) and then
```

File: utils/coders/decode.py [hunk 33]
```diff
@@ -28,10 +28,11 @@ def nSlotsFromParamsOutFile(filename):
 
 def parsePtxtGetSlots(iterable):
     for ptxt in iterable:
-        for slot in ast.literal_eval(ptxt):
+        for slot in ptxt:
             yield slot
 
 def printoutNums(ptxts, nelements):
+    breakpoint
```
Comment (jlhcrawford) [inline utils/coders/decode.py:35]:
```suggestion
```

File: include/helib/keys.h [hunk 34]
```diff
@@ -73,8 +70,14 @@ class PubKey
   Ctxt recryptEkey;  // the key itself, encrypted under key #0
 
 public:
+  /**
+   * @brief Class label to be added to JSON serialization as object type
+   * information.
+   */
+  static constexpr std::string_view typeName = "PubKey";
+
   //! This constructor thorws run-time error if activeContext=nullptr
```
Comment (hamishun) [inline include/helib/keys.h:79]:
```suggestion
```

File: src/binio.h [hunk 35]
```diff
@@ -0,0 +1,235 @@
...
+// or little endian.
+template <typename T>
+struct SerializeHeader
+{
+  // Header eye catcher
+  const std::array<char, EyeCatcher::SIZE> beginCatcher =
+      EyeCatcher::HEADER_BEGIN;
+  // 32 bit number: 8 bits for major, minor, patch, fix
+  const std::array<char, 4> version = Binio::VERSION_0_0_1_0;
+  //
```
Comment (hamishun) [inline src/binio.h:98]:
```suggestion
  // The helib version that output this header.
```

File: src/keys.cpp [hunk 36]
```diff
@@ -79,11 +81,11 @@ double RLWE(DoubleCRT& c0,
...
-PubKey::PubKey() :
-    context(*activeContext), pubEncrKey(*this), recryptEkey(*this)
-{
-  recryptKeyID = -1;
-}
+// PubKey::PubKey() :
+// context(*activeContext), pubEncrKey(*this), recryptEkey(*this)
+// {
+// recryptKeyID = -1;
+// }
```
Comment (hamishun) [inline src/keys.cpp:88]:
```suggestion
```

File: src/primeChain.cpp (hunk 30 above)
Comment (hamishun) [inline src/primeChain.cpp:35]:
```suggestion
```

File: src/primeChain.cpp [hunk 37]
```diff
@@ -19,45 +19,57 @@
...
 {
-  return s << '[' << e.first << ' ' << e.second << "]\n";
+  executeRedirectJsonError<void>([&]() {
+    json j{{"first", e.first}, {"second", unwrap(e.second.writeToJSON())}};
+    s << j.dump() << std::endl;
+  });
+  return s;
 }
+
+// FIXME: ADD JSON SERIALIZATION
```
Comment (hamishun) [inline src/primeChain.cpp:45]:
```suggestion
```

**Comment by hamishun** [comment]
Jenkins run tests

File: src/keys.cpp (hunk 29 above)
Comment (esteffin) [reply src/keys.cpp:1024]: yes

**Review by esteffin — APPROVED** [review]
Good with the exception of the 2 comments

**Comment by hamishun** [comment]
Jenkins run tests

**Comment by esteffin** [comment]
Jenkins run tests

**Comment by hamishun** [comment]
Jenkins run tests

**Comment by hamishun** [comment]
Jenkins run tests

**Review by faberga — APPROVED** [review]
Approved

**Pull request merged**